HTTP Cache

dantleech opened the pull request:
This PR introduces the `SuluHttpCacheBundle` and adds support for the Varnish cache proxy and invalidation through *tags*. Tags allow for very granular cache invalidation.

__Tasks:__

- [x] test coverage
- [ ] gather feedback for my changes
- [x] submit changes to the documentation
- [x] added changelog line
- [x] added to UPGRADE instructions
- [ ] ... <add your own tasks>


__Informations:__

| Q             | A
| ------------- | ---
| Tests pass?   | yes
| Fixed tickets | [ none
| BC Breaks     | 
| Doc           | https://github.com/sulu-cmf/sulu-docs/pull/8

dantleech commented on src/Sulu/Bundle/ContentBundle/Content/Types/SmartContent/SmartContent.php:
```diff
@@ -231,6 +231,21 @@ public function getContentData(PropertyInterface $property)
     /**
+     * {@inheritDoc}
+     */
+    public function getReferencedUuids(PropertyInterface $property)
```
x

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Compiler/HandlerPass.php:
```diff
@@ -0,0 +1,75 @@
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Add tagged cache handler definitions to the container
```
` ... to the aggregate cache handler`

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,107 @@
+
+        $this->configureStructureCacheHandlers($config['handlers'], $container);
+
+        if (isset($config['proxy_client'])) {
```
This should always be configured... remove the `if`

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // get default
+        $proxyClientName = null;
```
Move this block to configureProxyClient

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        $container->setAlias('sulu_http_cache.proxy_client', 'sulu_http_cache.proxy_client.' . $proxyClientName);
+        $container->setAlias('sulu_http_cache.default_handler', 'sulu_http_cache.handler.' . $config['default_handler']);
```
Move to handlers section

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+
+    public function configureStructureCacheHandlers($config, ContainerBuilder $container)
```
Docs ...

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/Resources/config/event-subscribers.xml:
```diff
@@ -0,0 +1,26 @@
+
+    <services>
+
+        <!-- event subscribers -->
```
Redundant.

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/Resources/config/proxy-client.xml:
```diff
@@ -0,0 +1,27 @@
+    </services>
+
+</container>
+
```
Extra line

dantleech commented on src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:
```diff
@@ -72,18 +72,6 @@ protected function renderStructure(
-
-                // set the private and shared max age
-                $response->setMaxAge(240);
-                $response->setSharedMaxAge(960);
```
Preview is not cached.

dantleech commented on src/Sulu/Component/HttpCache/EventSubscriber/ContentMapperSubscriber.php:
```diff
@@ -0,0 +1,53 @@
+     */
+    public function onContentNodeSave(ContentNodeEvent $event)
+    {
+        $this->handler->invalidateStructure($event->getStructure());
```
Yes we should...

dantleech replied:
Handled.

dantleech commented on src/Sulu/Component/HttpCache/Handler/AggregateHandler.php:
```diff
@@ -0,0 +1,51 @@
+use Symfony\Component\HttpFoundation\Response;
+use Sulu\Component\Content\StructureInterface;
+
+class AggregateHandler implements StructureCacheHandlerInterface
```
Doc

dantleech commented on src/Sulu/Component/HttpCache/Handler/AggregateHandler.php:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public function flush()
```
Doc (all this should be `{@inheritDoc}`

dantleech commented on src/Sulu/Component/HttpCache/Handler/StandardHandler.php:
```diff
@@ -0,0 +1,66 @@
+    {
+    }
+
+    public function updateResponse(Response $response, StructureInterface $structure)
```
doc

dantleech commented on src/Sulu/Component/HttpCache/StructureCacheHandlerInterface.php:
```diff
@@ -0,0 +1,15 @@
+
+    public function updateResponse(Response $response, StructureInterface $structure);
+
+    public function flush();
```
docs

dantleech commented on tests/Sulu/Component/HttpCache/Handler/AggregateHandlerTest.php:
```diff
@@ -0,0 +1,60 @@
+        $this->handler->updateResponse($this->response->reveal(), $this->structure->reveal());
+    }
+}
+
```
Extra line

dantleech commented on tests/Sulu/Component/HttpCache/Handler/TagsHandlerTest.php:
```diff
@@ -0,0 +1,84 @@
+
+        $this->handler->updateResponse($this->response->reveal(), $this->structure->reveal());
+    }
+
```
Extra line

dantleech commented:
Add Delete handling and KERNEL TERMINATE event handling (flush)

dantleech commented:
Add `supports` method to handlers .. check in aggregate handler

dantleech commented:
test flush's

dantleech commented:
Add debug info to response:

- Cache handlers
- Configured proxy client

dantleech commented on src/Sulu/Bundle/ContentBundle/Content/Types/SingleInternalLink.php:
```diff
@@ -42,12 +42,26 @@ public function write(
-            throw new \Exception();
+
+        if ($value === $node->getIdentifier()) {
+            throw new \InvalidArgumentException('Internal link node cannot reference itself');
```
Unrelated to this PR but fixes the above comment and .. well we should never throw an empty `Exception`.

dantleech commented on src/Sulu/Bundle/ContentBundle/Tests/Unit/Content/Types/InternalLinksTest.php:
```diff
@@ -0,0 +1,59 @@
+<?php
+
```
header

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,142 @@
+/**
+ * This is the class that loads and manages your bundle configuration
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html}
```
`{@inheritDoc}`

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,142 @@
+
+        $container->setParameter('sulu_http_cache.proxy_client.name', $proxyClientName);
+        $container->setAlias('sulu_http_cache.proxy_client', 'sulu_http_cache.proxy_client.' . $proxyClientName);
+
```
EXtra kue

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/SuluHttpCacheExtension.php:
```diff
@@ -0,0 +1,142 @@
+            if ($config['aggregate']['handlers']) {
+                $handlers = $config['aggregate']['handlers'];
+            } else {
+                $handlers = array_filter($enabledHandlers, function ($name) {
```
Add comment: Automatically add all the *enabled* handlers if no handlers are specified.

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/Resources/config/structure-cache-handlers.xml:
```diff
@@ -0,0 +1,50 @@
+        <service id="sulu_http_cache.handler.paths" class="%sulu_http_cache.handler.paths.class%">
+            <argument type="service" id="sulu_core.webspace.webspace_manager"/>
+            <argument type="service" id="sulu_http_cache.proxy_client"/>
+            <argument>%kernel.environment%</argument>
```
Is this required now?

dantleech replied:
Yes it is

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/SuluHttpCacheBundle.php:
```diff
@@ -0,0 +1,27 @@
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * SuluGeneratorBundle.
```
Wrong doc

dantleech commented on src/Sulu/Component/Content/Structure/Page.php:
```diff
@@ -16,11 +16,12 @@
  * This structure represents a page in the CMS
  */
-abstract class Page extends Structure
+abstract class Page extends Structure implements PageInterface
```
For some reason the Page was not implementing the `PageInterface`

dantleech commented:
This is now ready as far as I am concerned. See also the related documentation PR which is linked to in the description.

/cc @danrot @chirimoya

danrot commented on src/Sulu/Bundle/ContentBundle/Tests/Unit/Content/Types/SmartContent/SmartContentTest.php:
```diff
@@ -589,4 +573,51 @@ public function testGetViewDataPagedLimit($page, $pageSize, $offset, $limitResul
+
+    private function getContentDataProperty()
+    {
+        $property = $this->getMockForAbstractClass(
```
Definitely, that something I would say you can change as part of the boyscout rule ;-) But not really necessary in this PR.

danrot commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG for Sulu
 * dev-develop
     * FEATURE     #702 [AdminBundle]    Added sortings to user settings and changed default url for activities
     * BUGFIX      #697 [ContactBundle]  Set VAT number field optional
+    * FEATURE     #697 [ContactBundle]  Refactored HTTP cache, introduced Varnish support. See 38af8da73c929f9f57bb87a8973a1ee55dccee29
```
Should be content bundle, and line is probably too long

dantleech replied:
Should be `HttpCacheBundle/HttpCache`. The original code was in `CoreBundle` and the `HttpCache` component. It is mainly `HttpCacheBundle` though.

danrot commented on UPGRADE.md:
```diff
@@ -1,15 +1,40 @@
+dev-master
+----------
+
+The HTTP cache integration has been refactored. The following configuration
```
If you want to structure it like that we should probably introduce sub headlines, if you have multiple upgrade items it can be quite confusing to read otherwise, because of the missing structure.

dantleech replied:
ok. but I think we *have* to structure it like this, upgrade instructions cannot be implemented as a list.

danrot commented on UPGRADE.md:
```diff
@@ -1,15 +1,40 @@
-# Upgrade
+Upgrade
+=======
```
I would prefer the other variant of headlines, since we do it like that in all the markdown files, and it is easier to change the headline.

danrot commented on src/Sulu/Bundle/ContentBundle/Content/Types/SingleInternalLink.php:
```diff
@@ -42,12 +42,26 @@ public function write(
+            return array();
         }
+
+        return array($uuid);
```
I would use the ternary operator because of consistency with the `InternalLinks` content type.

dantleech replied:
+1

danrot commented on src/Sulu/Bundle/ContentBundle/Tests/Unit/Content/Types/InternalLinksTest.php:
```diff
@@ -0,0 +1,67 @@
+    private $contentQueryBuilder;
+    private $logger;
+    private $property;
+    private $type;
```
PHP Docs missing

danrot commented on src/Sulu/Bundle/ContentBundle/Tests/Unit/Content/Types/SingleInternalLinkTest.php:
```diff
@@ -0,0 +1,58 @@
+class SingleInternalLinkTest extends ProphecyTestCase
+{
+    private $property;
+    private $type;
```
missing phpdoc

danrot commented on src/Sulu/Component/HttpCache/Handler/AggregateHandler.php:
```diff
@@ -0,0 +1,111 @@
+
+namespace Sulu\Component\HttpCache\Handler;
+
+use Sulu\Component\HttpCache\StructureCacheHandlerInterface;
```
I can't find this class

danrot commented on src/Sulu/Bundle/SnippetBundle/Content/SnippetContent.php:
```diff
@@ -251,4 +252,23 @@ private function loadSnippets($ids, $webspaceKey, $locale, $shadowLocale = null)
+    private function getUuids($data)
+    {
+        $ids = is_array($data) && array_key_exists('ids', $data) ? $data['ids'] : array();
+        return $ids;
```
I don't really get that condition... What happens if the data is not an array?

dantleech replied:
I think its a bug in jackalope. Anyway, this is just moving the logic from one place to another, the contents of this is another issue.

danrot commented on src/Sulu/Bundle/TestBundle/Kernel/SuluTestKernel.php:
```diff
@@ -36,6 +36,8 @@ public function registerBundles()
             new \Sulu\Bundle\MediaBundle\SuluMediaBundle(),
             new \Sulu\Bundle\CategoryBundle\SuluCategoryBundle(),
+            new \Sulu\Bundle\HttpCacheBundle\SuluHttpCacheBundle(),
+            new \Sulu\Bundle\SnippetBundle\SuluSnippetBundle(),
```
It's not really a part of this PR, but I guess when we introduce the subtree split we have to be more concise here, otherwise all the bundles will depend on each other. Or we have to find a different solution at all :-)

danrot commented on src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:
```diff
@@ -67,26 +67,7 @@ protected function renderStructure(
-            $response->setContent($content);
-
-            // if not preview enable cache handling
-            if (!$preview) {
```
Where is the preview handled now?

danrot replied:
Nevermind, found it

danrot commented on src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:
```diff
@@ -67,26 +67,7 @@ protected function renderStructure(
-                // set reverse-proxy TTL (Symfony HttpCache, Varnish, ...)
-                $response->headers->set(
-                    HttpCache::HEADER_REVERSE_PROXY_TTL,
-                    $response->getAge() + intval($structure->getCacheLifeTime())
```
Is the structure cache lifetime still handled somewhere?

dantleech replied:
Yeah, in the `public` handler. See the documentation PR.

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/ContentMapperSubscriber.php:
```diff
@@ -0,0 +1,86 @@
+    public static function getSubscribedEvents()
+    {
+        return array(
+            ContentEvents::NODE_POST_SAVE => 'onContentNodeSave',
```
I guess this is not part of this PR, but shouldn't this event be called `onContentNodePostSave`?

danrot replied:
Ah, it is part of this PR... That's the method which should be called, right? Maybe we should call it the same as the event?

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriber.php:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        $this->handler->updateResponse($event->getResponse(), $this->structure);
+        $this->doFlush = true;
```
This value is never used, what is it intended for?

danrot commented on src/Sulu/Component/HttpCache/HandlerInterface.php:
```diff
@@ -0,0 +1,10 @@
+use Sulu\Component\Content\StructureInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+interface HandlerInterface
```
Does an empty standalone interface really make sense? Shouldn't the other interfaces inherit from it, so that there is at least some structure? Then all the Handlers would not have to additionally implement the HandlerInterface.

dantleech replied:
Yeah.. will make the other handler interfaces extend it.

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriber.php:
```diff
@@ -0,0 +1,111 @@
+     */
+    public function onTerminate(PostResponseEvent $event)
+    {
+        $this->handler->flush();
```
That will probably fail, if the passed Handler doesn't implement the HandlerFlushInterface, which is easily possible with you architecture. Wouldn't it be better to introduce `EventListeners`, and typehint for the correct interface in the constructor?

danrot replied:
Would probably make the `HandlerInterface` obsolete.

dantleech replied:
It still makes sense to have the HandlerInterface. Handlers may implement one or more of the Handler interfaces, but they are all handlers.

The validity of this design pattern used here is debatable, but its too late to change imo.

danrot replied:
It's never too late to change ;-) We are really giving up a lot of typing this way, which I don't like. I guess using EventListeners for this is not a really big change.

danrot commented on src/Sulu/Component/HttpCache/HandlerFlushInterface.php:
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
file header missing

danrot commented on src/Sulu/Component/HttpCache/HandlerFlushInterface.php:
```diff
@@ -0,0 +1,19 @@
+use Sulu\Component\Content\StructureInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+interface HandlerFlushInterface
```
phpdoc missing

danrot commented on src/Sulu/Component/HttpCache/HandlerInvalidateStructureInterface.php:
```diff
@@ -0,0 +1,16 @@
+<?php
+
```
file header missing

danrot commented on src/Sulu/Component/HttpCache/HandlerInvalidateStructureInterface.php:
```diff
@@ -0,0 +1,16 @@
+
+use Sulu\Component\Content\StructureInterface;
+
+interface HandlerInvalidateStructureInterface
```
phpdoc missing

danrot commented on src/Sulu/Component/HttpCache/HandlerUpdateResponseInterface.php:
```diff
@@ -0,0 +1,19 @@
+    public function updateResponse(Response $response, StructureInterface $structure);
+}
+
+
```
two empty lines

danrot commented on src/Sulu/Component/HttpCache/HandlerUpdateResponseInterface.php:
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
file header missing

danrot commented on src/Sulu/Component/HttpCache/HandlerUpdateResponseInterface.php:
```diff
@@ -0,0 +1,19 @@
+use Sulu\Component\Content\StructureInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+interface HandlerUpdateResponseInterface
```
phpdoc missing

danrot commented on src/Sulu/Component/HttpCache/HttpCache.php:
```diff
@@ -93,6 +93,6 @@ private function getReverseProxyTtl(Response $response)
         }
 
-        return null;
+        return;
```
would prefer `return null;`, since it is more explicit. And my IDE is complaining about it ;-)

dantleech replied:
Symfony code base: 321 `return;`, 136 `return null`. I see it as exiting the function which is not expected to have a return value.

dantleech replied:
Of those 136 the ones I checked are documented return values for functions which would could return a value. If the value is not intended to be used we should use `return;` I think. /cc @chirimoya

dantleech replied:
However in this case it should indeed be `null` as this method does return a value.

danrot commented on tests/Sulu/Component/HttpCache/EventSubscriber/ContentMapperSubscriberTest.php:
```diff
@@ -0,0 +1,53 @@
+    private $subscriber;
+    private $handler;
+    private $structure;
+    private $event;
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriberTest.php:
```diff
@@ -0,0 +1,115 @@
+    private $handler;
+    private $structure;
+    private $getResponseEvent;
+    private $filterResponseEvent;
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/Handler/AggregateHandlerTest.php:
```diff
@@ -0,0 +1,65 @@
+    private $handler1;
+    private $handler2;
+    private $structure;
+    private $response;
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/Handler/DebugHandlerTest.php:
```diff
@@ -0,0 +1,63 @@
+    private $handler;
+    private $handlerNames;
+    private $proxyClientName;
+    private $parameterBag;
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/Handler/PathsHandlerTest.php:
```diff
@@ -0,0 +1,86 @@
+    private $structure;
+    private $env = 'prod';
+    private $languageCode = 'en';
+    private $webspaceKey = 'sulu_io';
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/Handler/PublicHandlerTest.php:
```diff
@@ -0,0 +1,50 @@
+    private $structure;
+    private $maxAge = 10;
+    private $sharedMaxAge = 10;
+    private $parameterBag;
```
missing phpdocs

danrot commented on tests/Sulu/Component/HttpCache/Handler/TagsHandlerTest.php:
```diff
@@ -0,0 +1,86 @@
+class TagsHandlerTest extends ProphecyTestCase
+{
+    private $handler;
+    private $structure;
```
missing phpdocs

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriber.php:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * Get the Structure from the main request for use later.
+     *
+     * @param ContentNodeEvent $event
```
Wrong phpdoc

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/KernelSubscriber.php:
```diff
@@ -0,0 +1,111 @@
+    private $handler;
+
+    /**
+     * @var Sulu\Component\Content\StructureInterface
```
Generally I would prefer to use also the phpdocs, so that you don't have to put the FQN here

dantleech replied:
I don't think that would be accepted by the phpcsfixer and it is not the current practice in Symfony.

danrot replied:
But that's what we do all the time ;-)

dantleech replied:
But the `use` is for the compiler right? `use` for PHP documentor comments seems wrong to me.

dantleech replied:
ok. i think the general consensus is with you here. I think some of my doubts are invalid also as `use` is used regularaly with annotations (Doctrine, JMSSerializer, etc) and `phpcsfixer` does not destroy them :)

dantleech commented:
@danrot updated. if you want to merge, ping me and I will rebase.

danrot commented on CHANGELOG.md:
```diff
@@ -7,9 +7,11 @@ CHANGELOG for Sulu
 
 * 0.14.1 (2015-01-21)
     * HOTFIX      #737 [MediaBundle]    Changed BaseCollection properties to be protected for inheritance
+    * FEATURE     #697 [ContactBundle]  Refactored HTTP cache, introduced Varnish support. See 38af8da73c929f9f57bb87a8973a1ee55dccee29
```
Shouldn't this be removed?

dantleech replied:
..yeah

danrot commented on UPGRADE.md:
```diff
@@ -1,15 +1,35 @@
 
+## dev-master
+
+The HTTP cache integration has been refactored. The following configuration
```
Still can't see sub headlines ;-)

danrot commented on src/Sulu/Component/HttpCache/Handler/AggregateHandler.php:
```diff
@@ -0,0 +1,110 @@
+<?php
+
```
file header missing

dantleech commented:
Ready, depends also on  https://github.com/sulu-cmf/sulu-standard/pull/391

danrot commented:
Please remove the `HandlerInterface`, if it is still working you can merge it afterwards.

dantleech commented on tests/Sulu/Component/HttpCache/EventSubscriber/ContentMapperSubscriberTest.php:
```diff
@@ -0,0 +1,71 @@
+    private $subscriber;
+
+    /**
+     * @var HandlerInterface
```
@danrot this is an example of why we need the `HandlerInterface`. It is neither `FlushHandler`, `UpdateResponseHandler` or `StructureHandler`. it is all three. So we cannot identify it with a single type.

dantleech replied:
i.e. I can't really say `@var FlushHandler|UpdateResponseHandler|StructureHandler`

dantleech replied:
https://github.com/sulu-cmf/sulu/pull/653/files#diff-16ccae90fac80c28a8bfb96018750928R58 is a better example as this handler only implements one interface

danrot replied:
Still not a good reason, if you put HandlerInterface there you still get any helpful information

dantleech replied:
I disagree, from that you can detmerine which interfaces it could be, it defines the scope of the variable.

danrot replied:
Anyway, if i check the ContentMapperSubscriber you are using one of the concrete Interfaces

dantleech replied:
That may be, but the point is that this handler implements distinct 3 interfaces.
For this architecture to *make sense* you need an interface which ties them together.

How would you document the type of such an object? The only possiblity is `mixed` which implies that
it could be anything. If the type is `HandlerInterface` then you (or some static analysis) could infer
that it is one of the subtypes of that interface.

dantleech replied:
This is just an example and in the real code we don't *need* this, but these test cases illustrate the validity of a common handler interface. Otherwise I would have been happy to remove it.

danrot commented on src/Sulu/Component/HttpCache/EventSubscriber/ContentMapperSubscriber.php:
```diff
@@ -0,0 +1,86 @@
+class ContentMapperSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var Sulu\Component\HttpCache\StructureCacheHandlerInterface
```
This interface doesn't exist, please remove all of these references.

dantleech commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Compiler/HandlerPass.php:
```diff
@@ -0,0 +1,99 @@
+
+        if (!$reflection->implementsInterface('Sulu\Component\HttpCache\HandlerInterface')) {
+            throw new \InvalidArgumentException(sprintf(
+                'Service ID "%s" was tagged as a cache handler, but it does not implement the "HandlerInterface"'
```
Added compile time check for the HandlerInterface.

dantleech commented:
Removed and then re-added the HandlerInterface, After thinking about it, it is better to have it than not to have it.

With the handler interface we can ensure that the user has not made an error when tagging handler services. Without this interface the user can tag *any* service as a tag handler. This would be a silent mistake but would indicate the user has done something they did not intend to do.

The other argument is with auto-generated documentation. With this interface it is easy to see which handlers exist. It is a form of categorization.

With this in mind I have added a compile-time check to the dependency injection builder to ensure that the classes being added to the AggregateHandler implement the `HandlerInterface`.

danrot commented:
Now you can still implement a class like the following, which should not be a valid handler IMO:

```php
<?php
class SomeClass implements HandlerInterface {}
```

But at least you have extended this interface now, so it is ok for me now.

danrot commented on src/Sulu/Bundle/HttpCacheBundle/DependencyInjection/Compiler/HandlerPass.php:
```diff
@@ -0,0 +1,99 @@
+        $definition = $container->getDefinition($id);
+        $reflection = new \ReflectionClass($definition->getClass());
+
+        if (!$reflection->implementsInterface('Sulu\Component\HttpCache\HandlerInterface')) {
```
AFAIK `instanceof` is faster than using Reflections.

dantleech replied:
yeah I believe it is, but you can only do instanceof on instantiated classes. also performance is a non-issue here as this is only executed when the container is being built. (its then cached even in dev mode)

danrot replied:
True, then it is fine.
